src/transport/Tuple.h

```diff
@@ -0,0 +1,215 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ * @file
+ *
+ * Defines a compound transport type (tuple) that can merge several transports
+ * to look like a single one.
+ */
+#ifndef TRANSPORT_TUPLE_H_
+#define TRANSPORT_TUPLE_H_
+
+#include <tuple>
+#include <type_traits>
+
+#include <transport/Base.h>
+
+namespace chip {
+namespace Transport {
+
+/**
+ * Groups together several transports of different types and presents them as a unified transport.
+ *
+ * The usage intent is to be able to group several distinct transport types and make them look
+ * as a single transport.
+ *
+ * Having an example class definition of:
+ *    Tuple<UDP, UDP, TCP>
+ *
+ * Is equivalent to:
+ *    class TupleOfUdpUdpTcp : public BASE {
+ *       private:
+ *          UDP mUdp1;
+ *          UDP mUdp2;
+ *          TCP mTcp2;
+ *       public:
+ *          Init(UDPListenParameters &, UDPListenParameters&, TCPListenParameters) {...}
+ *          CHIP_ERROR SendMessage(...) override;
+ *          bool CanSendToPeer(...) override;
+ *    }
+ *
+ * The intent of this is to allow applications to use any transport types without CHIP pre-defining
+ * popular mappings (like UDP only, UDP and BLE, BLE only etc.) and without using #ifdefs to create
+ * a single 'standard transport'.
+ *
+ * Transport logic:
+ *   - Every transport can decide if a 'PeerAddress' can be sent over 'self'
+ *
+ *   - Within a mixed tuple, if several transports support a peer address (e.g. TCP and UDP both
+ *     support IP), the first one found wins.
+ *
+ *   - Expected message sending logic:
+ *     - BLE transports only support BLE. TCP/UDP support IP addresses
+ *     - Multicasts only supported by UDP
+ *
+ * @param TransporTypes the ORDERED list of transport types grouped together. Order matters in
```

Comment:
I think `@param` needs to be `@tparam`, per [here](https://www.doxygen.nl/manual/commands.html#cmdtparam).

Also `TransporTypes` -> `TransportTypes`.

Reply:
There are other template parameter instances below that I believe need to be likewise tweaked.

Reply:
Updated. VSCode colors `@param` but not `@tparam` so I was unsure if the tparam is a thing or not. A search seems to indicate it exists, so I will trust it.